### ISSUE TITLE
how can I authenticate proxy? I will be grateful for the answer.

### ISSUE THREAD

**Issue by webjohny** [chromedp/chromedp#645]
#### What versions are you running?

<pre>
github.com/chromedp/chromedp v0.5.3
83.0.4103.106
go version go1.14.1 windows/amd64
</pre>

#### What did you do? Include clear steps.
I need to authenticate a proxy server without my intervention.
I tried to make analog the function from puppeteer (page.authenticate), but I didn't catch event Fetch.authRequired.

<pre>
err := network.SetCacheDisabled(true).Do(ctx)
fmt.Println(err)

pattern := fetch.RequestPattern{}
pattern.URLPattern = "*"

reqPatterns := []*fetch.RequestPattern{}
reqPatterns = append(reqPatterns, &pattern)

fetchEnabled := fetch.Enable().WithPatterns(reqPatterns).WithHandleAuthRequests(true)
err = fetchEnabled.Do(ctx)
</pre>

#### What did you expect to see?
I want to catch event Fetch.authRequired so that continue to connect authentication func. 

#### What did you see instead?
I cannot understand how to make this analog of puppeteer function

**Comment by webjohny** [comment]
I did it)
It has problem with events queue and you cannot send ContinueRequest to between "Fetch.enable" delay and "Fetch.continueRequest" and you have to write code where will be sending the request "Fetch.continueRequest" without delays. 
Steps: 
1. needs to fork this repo
2. change property "conn Transport" in Browser struct => "Conn Transport" so that you can call it from your script
3. next in your script you have to call "Fetch.enable"
<pre>
pattern := fetch.RequestPattern{}
pattern.URLPattern = "*"

var reqPatterns []*fetch.RequestPattern
reqPatterns = append(reqPatterns, &pattern)

fetchEnabled := fetch.Enable().WithPatterns(reqPatterns).WithHandleAuthRequests(true)

if err := chromedtp.Run(j.ctx,
        network.Enable(),
	performance.Enable(),
	page.SetLifecycleEventsEnabled(true),
	security.SetIgnoreCertificateErrors(true),
	emulation.SetTouchEmulationEnabled(false),
	network.SetCacheDisabled(true),
        fetchEnabled,
        chromedtp.ActionFunc(func (ctx context.Context) error {
		j.ListenForNetworkEvent(ctx)
		return nil
	}),
	chromedtp.Navigate("https://myip.ru"),
); err != nil {
	log.Fatal(err)
}
</pre>
4. and then can chromedp.Navigate("your link")
5. next you have to write ListenTarget where you catch event "Fetch.requestPaused" and there need write handler
<pre>
chromedtp.ListenTarget(ctx, func(ev interface{}) {
	switch ev := ev.(type) {

	case *fetch.EventAuthRequired:
		fmt.Println("*fetch.EventAuthRequired", ev.RequestID)
		c := chromedtp.FromContext(ctx)
		buf, _ := json.Marshal(map[string]interface{}{
			"requestId": ev.RequestID.String(),
			"authChallengeResponse": map[string]string{
				"response": "ProvideCredentials",
				"username": j.proxy.Login,
				"password": j.proxy.Password,
			},
		})
		cmd := &cdproto.Message{
			ID:        20,
			SessionID: c.Target.SessionID,
			Method:    cdproto.MethodType("Fetch.continueWithAuth"),
			Params:    buf,
		}
		err := c.Browser.Conn.Write(ctx, cmd)

	case *fetch.EventRequestPaused:
		j.interceptionID = ev.RequestID
		c := chromedtp.FromContext(ctx)
		buf, _ := json.Marshal(map[string]string{"requestId":ev.RequestID.String()})
		cmd := &cdproto.Message{
			ID:        20,
			SessionID: c.Target.SessionID,
			Method:    cdproto.MethodType("Fetch.continueRequest"),
			Params:    buf,
		}
		err := c.Browser.Conn.Write(ctx, cmd)
		fmt.Println(err)
	}
})
</pre>
6. Finally it will be working)

**Comment by ysmood** [comment]
@webjohny This may help you https://github.com/go-rod/rod/issues/45

**Comment by fredipevcin** [comment]
@webjohny would this help? Code is missing checks if the authentication is for proxy among other things =) but to give you a bit start.
```go
package main

import (
	"context"
	"fmt"
	"log"
	"time"

	"github.com/chromedp/cdproto/cdp"
	"github.com/chromedp/cdproto/fetch"
	"github.com/chromedp/chromedp"
)

func main() {
	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
	defer cancel()
	ctx, cancel = chromedp.NewExecAllocator(
		ctx,
		append(chromedp.DefaultExecAllocatorOptions[:], chromedp.ProxyServer("http://proxyip:port"))...,
	)
	defer cancel()
	ctx, cancel = chromedp.NewContext(ctx)
	defer cancel()
	defer chromedp.Cancel(ctx)

	chromedp.ListenTarget(ctx, func(ev interface{}) {
		go func() {
			switch ev := ev.(type) {
			case *fetch.EventAuthRequired:
				c := chromedp.FromContext(ctx)
				execCtx := cdp.WithExecutor(ctx, c.Target)

				resp := &fetch.AuthChallengeResponse{
					Response: fetch.AuthChallengeResponseResponseProvideCredentials,
					Username: "proxy username",
					Password: "proxy password",
				}

				err := fetch.ContinueWithAuth(ev.RequestID, resp).Do(execCtx)
				if err != nil {
					log.Print(err)
				}

			case *fetch.EventRequestPaused:
				c := chromedp.FromContext(ctx)
				execCtx := cdp.WithExecutor(ctx, c.Target)
				err := fetch.ContinueRequest(ev.RequestID).Do(execCtx)
				if err != nil {
					log.Print(err)
				}
			}
		}()
	})

	var html string
	if err := chromedp.Run(ctx,
		fetch.Enable().WithHandleAuthRequests(true),
		chromedp.Navigate("http://api.ipify.org?format=json"),
		chromedp.OuterHTML("html", &html),
	); err != nil {
		log.Fatal(err)
	}

	fmt.Println(html)
}
```

**Comment by bostin** [comment]
> @webjohny would this help? Code is missing checks if the authentication is for proxy among other things =) but to give you a bit start.
> 
> ```go
> package main
> 
> import (
> 	"context"
> 	"fmt"
> 	"log"
> 	"time"
> 
> 	"github.com/chromedp/cdproto/cdp"
> 	"github.com/chromedp/cdproto/fetch"
> 	"github.com/chromedp/chromedp"
> )
> 
> func main() {
> 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
> 	defer cancel()
> 	ctx, cancel = chromedp.NewExecAllocator(
> 		ctx,
> 		append(chromedp.DefaultExecAllocatorOptions[:], chromedp.ProxyServer("http://proxyip:port"))...,
> 	)
> 	defer cancel()
> 	ctx, cancel = chromedp.NewContext(ctx)
> 	defer cancel()
> 	defer chromedp.Cancel(ctx)
> 
> 	chromedp.ListenTarget(ctx, func(ev interface{}) {
> 		go func() {
> 			switch ev := ev.(type) {
> 			case *fetch.EventAuthRequired:
> 				c := chromedp.FromContext(ctx)
> 				execCtx := cdp.WithExecutor(ctx, c.Target)
> 
> 				resp := &fetch.AuthChallengeResponse{
> 					Response: fetch.AuthChallengeResponseResponseProvideCredentials,
> 					Username: "proxy username",
> 					Password: "proxy password",
> 				}
> 
> 				err := fetch.ContinueWithAuth(ev.RequestID, resp).Do(execCtx)
> 				if err != nil {
> 					log.Print(err)
> 				}
> 
> 			case *fetch.EventRequestPaused:
> 				c := chromedp.FromContext(ctx)
> 				execCtx := cdp.WithExecutor(ctx, c.Target)
> 				err := fetch.ContinueRequest(ev.RequestID).Do(execCtx)
> 				if err != nil {
> 					log.Print(err)
> 				}
> 			}
> 		}()
> 	})
> 
> 	var html string
> 	if err := chromedp.Run(ctx,
> 		fetch.Enable().WithHandleAuthRequests(true),
> 		chromedp.Navigate("http://api.ipify.org?format=json"),
> 		chromedp.OuterHTML("html", &html),
> 	); err != nil {
> 		log.Fatal(err)
> 	}
> 
> 	fmt.Println(html)
> }
> ```

It's working, thanks

**Comment by ZekeLu** [comment]
A PR (chromedp/examples#30) based on @fredipevcin's comment has been created. Thank you @fredipevcin!

The differences in the PR are:

1. it does not create a goroutine for every event;
2. it disables the `Fetch` domain and removes the event handler once the proxy is authenticated (to reduce overhead);
3. it uses `chromedp.Run` in the event handler to simplify the code.

I will close this issue once the PR is merged.

**Comment by whiteCcinn** [comment]
> @webjohny这有帮助吗？
> 
> ```go
> package main
> 
> import (
> 	"context"
> 	"fmt"
> 	"log"
> 	"time"
> 
> 	"github.com/chromedp/cdproto/cdp"
> 	"github.com/chromedp/cdproto/fetch"
> 	"github.com/chromedp/chromedp"
> )
> 
> func main() {
> 	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
> 	defer cancel()
> 	ctx, cancel = chromedp.NewExecAllocator(
> 		ctx,
> 		append(chromedp.DefaultExecAllocatorOptions[:], chromedp.ProxyServer("http://proxyip:port"))...,
> 	)
> 	defer cancel()
> 	ctx, cancel = chromedp.NewContext(ctx)
> 	defer cancel()
> 	defer chromedp.Cancel(ctx)
> 
> 	chromedp.ListenTarget(ctx, func(ev interface{}) {
> 		go func() {
> 			switch ev := ev.(type) {
> 			case *fetch.EventAuthRequired:
> 				c := chromedp.FromContext(ctx)
> 				execCtx := cdp.WithExecutor(ctx, c.Target)
> 
> 				resp := &fetch.AuthChallengeResponse{
> 					Response: fetch.AuthChallengeResponseResponseProvideCredentials,
> 					Username: "proxy username",
> 					Password: "proxy password",
> 				}
> 
> 				err := fetch.ContinueWithAuth(ev.RequestID, resp).Do(execCtx)
> 				if err != nil {
> 					log.Print(err)
> 				}
> 
> 			case *fetch.EventRequestPaused:
> 				c := chromedp.FromContext(ctx)
> 				execCtx := cdp.WithExecutor(ctx, c.Target)
> 				err := fetch.ContinueRequest(ev.RequestID).Do(execCtx)
> 				if err != nil {
> 					log.Print(err)
> 				}
> 			}
> 		}()
> 	})
> 
> 	var html string
> 	if err := chromedp.Run(ctx,
> 		fetch.Enable().WithHandleAuthRequests(true),
> 		chromedp.Navigate("http://api.ipify.org?format=json"),
> 		chromedp.OuterHTML("html", &html),
> 	); err != nil {
> 		log.Fatal(err)
> 	}
> 
> 	fmt.Println(html)
> }
> ```

Thank you very much. It worked for me. It was great.

**Comment by Rajkumar0718** [comment]
fetch.EventAuthRequired not triggering while i am giving handleAuthRequests as true in fetch.enable method

**Comment by ZekeLu** [comment]
@Rajkumar0718 You have commented on #851 before. Please file a new issue and provide a self-contained demo to show the issue. Thanks!
